File: 2023/day01/src/main.rs
use anyhow::Result;
use std::collections::HashMap;
use std::env;
use std::fs::File;
use std::io::{BufRead, BufReader};

fn main() -> Result<()> {
    let args = env::args().nth(1).expect("Please provide a file name");
    let file = File::open(args)?;
    let reader = BufReader::new(file);
    let mut total1 = 0;
    let mut total2 = 0;
    let numbers: HashMap<&str, u32> = HashMap::from([
        ("one", 1),
        ("two", 2),
        ("three", 3),
        ("four", 4),
        ("five", 5),
        ("six", 6),
        ("seven", 7),
        ("eight", 8),
        ("nine", 9),
    ]);
    let mut found: bool;
    for line in reader.lines() {
        let line = line?;
        let start = line.chars().position(|c| c.is_ascii_digit()).unwrap();
        let x1 = line.chars().nth(start).unwrap().to_digit(10).unwrap();
        let mut x2 = x1;
        found = false;
        for i in 1..start + 1 {
            for (k, v) in numbers.iter() {
                if line[..i].contains(k) {
                    x2 = *v;
                    found = true;
                    break;
                }
            }
            if found {
                break;
            }
        }
        let end = line.len() - line.chars().rev().position(|c| c.is_ascii_digit()).unwrap() - 1;
        let y1 = line.chars().nth(end).unwrap().to_digit(10).unwrap();
        let mut y2 = y1;
        let mut i = line.len() - 2;
        found = false;
        while i > end && !found {
            for (k, v) in numbers.iter() {
                if line[i..].contains(k) {
                    y2 = *v;
                    found = true;
                    break;
                }
            }
            i -= 1;
        }

        total1 += 10 * x1 + y1;
        total2 += 10 * x2 + y2;
    }
    println!("Part 1: {}", total1);
    println!("Part 2: {}", total2);
    Ok(())
}

File: 2023/day01/main.c
#include "stdio.h"
#include "stdlib.h"
#include "string.h"

#define BUFFER_SIZE 256

int main(int argc, char *argv[]) {
  if (argc < 2) {
    printf("Missing input file\n");
    return 1;
  }

  char *numbers[9] = {"one", "two",   "three", "four", "five",
                      "six", "seven", "eight", "nine"};
  int values[9] = {1, 2, 3, 4, 5, 6, 7, 8, 9};

  FILE *file = fopen(argv[1], "r");
  int part1 = 0, part2 = 0;
  char line[BUFFER_SIZE];
  while (fgets(line, BUFFER_SIZE, file)) {
    int n = strcspn(line, "\n");
    line[n] = '\0';
    int start = 0;
    while (line[start] < '1' || line[start] > '9')
      start++;
    int x1 = line[start] - '0';
    int x2 = x1;
    int found = 0;
    for (int i = 1; i < start && !found; ++i) {
      char slice[i + 1];
      strncpy(slice, line, i + 1);
      for (int j = 0; j < 9; ++j) {
        if (strstr(slice, numbers[j])) {
          x2 = values[j];
          found = 1;
          break;
        }
      }
    }
    int end = n - 1;
    while (line[end] < '1' || line[end] > '9')
      end--;
    int y1 = line[end] - '0';
    int y2 = y1;
    found = 0;
    for (int i = n - 2; i > end && !found; --i) {
      for (int j = 0; j < 9; ++j) {
        if (strstr(line + i, numbers[j])) {
          y2 = values[j];
          found = 1;
          break;
        }
      }
    }
    part1 += x1 * 10 + y1;
    part2 += x2 * 10 + y2;
  }
  fclose(file);

  printf("Part 1: %d\n", part1);
  printf("Part 2: %d\n", part2);

  return 0;
}
